Proposal for common Cloud Controller Manager repository structure

hogepodge opened the pull request:
This document defines a common code location and layout for provider
Cloud Controller Manager implementations.

https://github.com/kubernetes/community/pull/1862

hogepodge commented:
wg/cloud-provider
sig/aws
sig/azure
sig/gce
sig/openstack

sig/docs
sig/testing

feiskyer commented:
@hogepodge  Will someone create repos with this structure for all cloud providers?

hogepodge commented:
@feiskyer yes, @cheftako knows who we need to ask to create the repos.

Looking back at the notes from our meeting, I think I need to expand the requirements list to include:

* ccm goes into the ccm dir
* dns goes into dns dir
* cni driver goes into cni dir
* flex volume driver goes into flex dir
* installer goes into installer dir
* tests goes into tests dir

wlan0 commented:
@hogepodge I believe @calebamiles can help with the creation of these repos in the org.

> ccm goes into the ccm dir
dns goes into dns dir
cni driver goes into cni dir
flex volume driver goes into flex dir
installer goes into installer dir
tests goes into tests dir

Pls tag me for approval once you update the doc to reflect the above directory structure.

wlan0 commented on wg-cloud-provider/cloud-provider-requirements.md:
```diff
@@ -0,0 +1,67 @@
+## Requirements
+
+Each cloud provider hosted within the `kubernetes` organization shall have a
+single repository named `kubernetes/cloud-provider-<provider_name>`. Those
```
+1

yastij replied:
I agree with @wlan0

wlan0 commented on wg-cloud-provider/cloud-provider-requirements.md:
```diff
@@ -0,0 +1,67 @@
+provider repositories owned by current and future cloud provider SIGs[1]. In
+accordance with the WG-Cloud-Provider Charter[2] to "define a set of common
+expected behaviors across cloud providers", this proposal defines the location
+and structure of commonly expected code.
```
I'm not sure if we should add a sentence stating that this document only applies to cloud providers that have been approved by CNCF to be a part of the Kubernetes org. @calebamiles thoughts?

dims replied:
Where is that list @wlan0 ? What would a cloud provider have to do to get into that list?

stmcginnis replied:
I think it may be better to not call out any kind of official/approved status. I could see potentially causing issues with locating cloud providers and confusion for end users as to where the "official" version resides. Until a provider is accepted, it could cause duplication of multiple people working on the same thing. Better to have one common location where interested parties can all come together to collaborate, IMO.

wlan0 replied:
@dims https://github.com/kubernetes/community/pull/1712/files#diff-1d931bbfa82e30aa84cde953b196af78R305 

@hogepodge We'll need to communicate with @calebamiles so that there is no overlap of work. The PR above overlaps with some of the information in this PR.

> I could see potentially causing issues with locating cloud providers and confusion for end users as to where the "official" version resides.

@stmcginnis There is a well defined process to ensure that this does not happen. Please refer to the link pasted.

idvoretskyi replied:
@wlan0 

> I'm not sure if we should add a sentence stating that this document only applies to cloud providers that have been approved by CNCF to be a part of the Kubernetes org.

Not sure if I understood how CNCF can be influential here.

aoxn replied:
Hi @wlan0  ,   I had the same concern with @dims , The Alibaba Cloud also want to help with the out of tree cloud provider.  And what we should do to go into that list? Is there a standard procedure?  Thanks.

hogepodge commented:
@wlan0 updated

stmcginnis commented on wg-cloud-provider/cloud-provider-requirements.md:
```diff
@@ -0,0 +1,87 @@
+expected locations and naming conventions of where the external provider code
+is, we continue in creating a consistent experience for:
+
+* Users of the providers, who will have easily understandable conventions for
```
+1

xing-yang commented on wg-cloud-provider/cloud-provider-requirements.md:
```diff
@@ -0,0 +1,87 @@
+
+* `dns` for DNS provider code.
+* `cni` for the Container Network Interface (CNI) driver.
+* `flex` for the Flex Volume driver.
```
There should be a subdirectory for `csi`.

idvoretskyi commented:
PTAL @kubernetes/sig-architecture-pr-reviews

andrewsykim commented on wg-cloud-provider/cloud-provider-requirements.md:
```diff
@@ -0,0 +1,87 @@
+Kubernetes 1.11 development cycle, with conformance by the 1.11 release.
+
+After the 1.11 release all current and new provider implementations must
+conform with the requirements outlined in this document.
```
I think it's reasonable to assume that the repo structure we want may change (a lot) as we test it out. Having said that we should include in the timeline section that at any time before the 1.11 release we may make drastic changes.

wlan0 replied:
I agree with @andrewsykim. This LGTM as it is right now. We still need to add the `csi` sub-dir, but let's get this merged and then edit this as needed.

wlan0 commented:
/lgtm
/approve

wlan0 commented:
/assign @cblecker

cblecker commented:
/approve

hogepodge commented:
Please see follow up patch addressing some of the issues raised in the comments and wg-cloud-provider. https://github.com/kubernetes/community/pull/1942
